src/decoders/belief_propogation.jl

```diff
@@ -0,0 +1,152 @@
+using SparseArrays
+
+include("abstract_decoder.jl")
+
+struct BeliefPropagationSetup
```

Comment:
May we rename this to `BeliefPropagationScratchSpace` or `BeliefPropagationBuffer` or something similar?